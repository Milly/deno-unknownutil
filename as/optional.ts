import type { Predicate } from "../is.ts";
import { rewriteName } from "../_funcutil.ts";
import { annotate, hasAnnotation, unannotate } from "../_annotation.ts";

/**
 * Annotation for optional.
 */
export type WithOptional<T> = {
  optional: Predicate<T>;
};

/**
 * Return an `Optional` annotated type predicate function that returns `true` if the type of `x` is `T` or `undefined`.
 *
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.ObjectOf({
 *   foo: as.Optional(is.String),
 * });
 * const a: unknown = {};
 * if (isMyType(a)) {
 *   // a is narrowed to {foo?: string}
 *   const _: {foo?: string} = a;
 * }
 * ```
 */
export function asOptional<T>(
  pred: Predicate<T>,
): Predicate<T | undefined> & WithOptional<T> {
  if (hasAnnotation(pred, "optional")) {
    return pred as Predicate<T | undefined> & WithOptional<T>;
  }
  return rewriteName(
    annotate(
      (x: unknown): x is T | undefined => x === undefined || pred(x),
      "optional",
      pred,
    ),
    "asOptional",
    pred,
  ) as Predicate<T | undefined> & WithOptional<T>;
}

/**
 * Return an `Optional` un-annotated type predicate function that returns `true` if the type of `x` is `T`.
 *
 * To enhance performance, users are advised to cache the return value of this function and mitigate the creation cost.
 *
 * ```ts
 * import { as, is } from "@core/unknownutil";
 *
 * const isMyType = is.ObjectOf({
 *   foo: as.Unoptional(as.Optional(is.String)),
 * });
 * const a: unknown = {foo: "a"};
 * if (isMyType(a)) {
 *   // a is narrowed to {foo: string}
 *   const _: {foo: string} = a;
 * }
 * ```
 */
export function asUnoptional<
  P extends Predicate<unknown>,
  T extends P extends Predicate<infer T | undefined> ? T
    : P extends Predicate<infer T> ? T
    : never,
>(pred: P): Predicate<T> {
  if (!hasAnnotation(pred, "optional")) {
    return pred as Predicate<T>;
  }
  return unannotate(pred, "optional") as Predicate<T>;
}

/**
 * Check if the given type predicate has optional annotation.
 */
export function hasOptional<
  P extends Predicate<unknown>,
  T extends P extends Predicate<infer T | undefined> ? T
    : P extends Predicate<infer T> ? T
    : never,
>(
  pred: P,
): pred is P & WithOptional<T> {
  return hasAnnotation(pred, "optional");
}
