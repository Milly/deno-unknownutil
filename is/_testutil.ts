// It seems 'IsExact' in deno_std is false positive so use `Equal` in type-challenges
// https://github.com/type-challenges/type-challenges/blob/e77262dba62e9254451f661cb4fe5517ffd1d933/utils/index.d.ts#L7-L9
export type Equal<X, Y> = (<T>() => T extends X ? 1 : 2) extends
  (<T>() => T extends Y ? 1 : 2) ? true : false;

export function stringify(x: unknown): string {
  if (x instanceof Date) return `Date(${x.valueOf()})`;
  if (x instanceof Promise) return "Promise";
  if (typeof x === "function") return x.toString();
  if (typeof x === "bigint") return `${x}n`;
  if (typeof x === "symbol") return x.toString();
  return JSON.stringify(x);
}
