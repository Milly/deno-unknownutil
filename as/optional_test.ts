import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import { type Equal, testWithExamples } from "../_testutil.ts";
import { is } from "../is/mod.ts";
import { asOptional, asUnoptional } from "./optional.ts";

Deno.test("asOptional<T>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, asOptional(is.Number).name);
    // Nesting does nothing
    await assertSnapshot(t, asOptional(asOptional(is.Number)).name);
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = undefined;
    if (is.ObjectOf({ a: asOptional(is.Number) })(a)) {
      assertType<Equal<typeof a, { a?: number }>>(true);
    }
  });
  await t.step("with is.String", async (t) => {
    await testWithExamples(t, asOptional(is.String), {
      validExamples: ["string", "undefined"],
    });
  });
  await t.step("with is.Number", async (t) => {
    await testWithExamples(t, asOptional(is.Number), {
      validExamples: ["number", "undefined"],
    });
  });
  await t.step("with is.Bigint", async (t) => {
    await testWithExamples(t, asOptional(is.Bigint), {
      validExamples: ["bigint", "undefined"],
    });
  });
  await t.step("with is.Boolean", async (t) => {
    await testWithExamples(t, asOptional(is.Boolean), {
      validExamples: ["boolean", "undefined"],
    });
  });
  await t.step("with is.Array", async (t) => {
    await testWithExamples(t, asOptional(is.Array), {
      validExamples: ["array", "undefined"],
    });
  });
  await t.step("with is.Set", async (t) => {
    await testWithExamples(t, asOptional(is.Set), {
      validExamples: ["set", "undefined"],
    });
  });
  await t.step("with is.RecordObject", async (t) => {
    await testWithExamples(t, asOptional(is.RecordObject), {
      validExamples: ["record", "undefined"],
    });
  });
  await t.step("with is.Function", async (t) => {
    await testWithExamples(t, asOptional(is.Function), {
      validExamples: ["syncFunction", "asyncFunction", "undefined"],
    });
  });
  await t.step("with is.SyncFunction", async (t) => {
    await testWithExamples(t, asOptional(is.SyncFunction), {
      validExamples: ["syncFunction", "undefined"],
    });
  });
  await t.step("with is.AsyncFunction", async (t) => {
    await testWithExamples(t, asOptional(is.AsyncFunction), {
      validExamples: ["asyncFunction", "undefined"],
    });
  });
  await t.step("with is.Null", async (t) => {
    await testWithExamples(t, asOptional(is.Null), {
      validExamples: ["null", "undefined"],
    });
  });
  await t.step("with is.Undefined", async (t) => {
    await testWithExamples(t, asOptional(is.Undefined), {
      validExamples: ["undefined"],
    });
  });
  await t.step("with is.Symbol", async (t) => {
    await testWithExamples(t, asOptional(is.Symbol), {
      validExamples: ["symbol", "undefined"],
    });
  });
});

Deno.test("asUnoptional<T>", async (t) => {
  await t.step("returns properly named function", async (t) => {
    await assertSnapshot(t, asUnoptional(asOptional(is.Number)).name);
    // Non optional does nothing
    await assertSnapshot(t, asUnoptional(is.Number).name);
    // Nesting does nothing
    await assertSnapshot(
      t,
      asUnoptional(asUnoptional(asOptional(is.Number))).name,
    );
  });
  await t.step("returns proper type predicate", () => {
    const a: unknown = undefined;
    if (is.ObjectOf({ a: asUnoptional(asOptional(is.Number)) })(a)) {
      assertType<Equal<typeof a, { a: number }>>(true);
    }
    if (asUnoptional(is.Number)(a)) {
      assertType<Equal<typeof a, number>>(true);
    }
  });
  await t.step("with is.String", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.String)), {
      validExamples: ["string"],
    });
  });
  await t.step("with is.Number", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.Number)), {
      validExamples: ["number"],
    });
  });
  await t.step("with is.Bigint", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.Bigint)), {
      validExamples: ["bigint"],
    });
  });
  await t.step("with is.Boolean", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.Boolean)), {
      validExamples: ["boolean"],
    });
  });
  await t.step("with is.Array", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.Array)), {
      validExamples: ["array"],
    });
  });
  await t.step("with is.Set", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.Set)), {
      validExamples: ["set"],
    });
  });
  await t.step("with is.RecordObject", async (t) => {
    await testWithExamples(
      t,
      asUnoptional(asOptional(is.RecordObject)),
      {
        validExamples: ["record"],
      },
    );
  });
  await t.step("with is.Function", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.Function)), {
      validExamples: ["syncFunction", "asyncFunction"],
    });
  });
  await t.step("with is.SyncFunction", async (t) => {
    await testWithExamples(
      t,
      asUnoptional(asOptional(is.SyncFunction)),
      {
        validExamples: ["syncFunction"],
      },
    );
  });
  await t.step("with is.AsyncFunction", async (t) => {
    await testWithExamples(
      t,
      asUnoptional(asOptional(is.AsyncFunction)),
      {
        validExamples: ["asyncFunction"],
      },
    );
  });
  await t.step("with is.Null", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.Null)), {
      validExamples: ["null"],
    });
  });
  await t.step("with is.Undefined", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.Undefined)), {
      validExamples: ["undefined"],
    });
  });
  await t.step("with is.Symbol", async (t) => {
    await testWithExamples(t, asUnoptional(asOptional(is.Symbol)), {
      validExamples: ["symbol"],
    });
  });
});
