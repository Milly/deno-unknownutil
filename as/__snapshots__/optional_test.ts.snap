export const snapshot = {};

snapshot[`asUnoptional<T> > returns properly named function 1`] = `"isNumber"`;

snapshot[`asUnoptional<T> > returns properly named function 2`] = `"isNumber"`;

snapshot[`asUnoptional<T> > returns properly named function 3`] = `"isNumber"`;

snapshot[`asOptional<T> > returns properly named function 1`] = `"asOptional(isNumber)"`;

snapshot[`asOptional<T> > returns properly named function 2`] = `"asOptional(isNumber)"`;
